0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+0. The sender encrypts the inner headers and their data with the payment key using AES-256-CTR.
+0. The sender creates a PSK packet containing the outer (public) headers, and attaches the token and HMAC function generated the payment key with as headers. The encrypted inner headers are attached as binary data.
+0. The sender quotes to the receiver and sends a transfer with the outer headers in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
+0. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment key from the outer headers contained in the data field of the ILP packet, and uses it to decrypt the inner headers and their data.
+0. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two sets of headers, like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Custom binary data is appended to the inner headers, after a blank line. This object is then encrypted and appended to the outer headers as binary data, after a blank line. Both inner and outer headers are parsed in the exact same way as HTTP headers.
+
+The outer headers come after a single status line, which must be `PSK/1.0`. The outer headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner headers.
+
+### Outer Request
+
+The outer headers and their data are formatted as below:
+
+```http
+PSK/1.0 PUBLIC
+Key: hmac-sha-256 fpwpAhlN588
+Header: data_everyone_can_see
+
+...
+```
+
+Everything stored in these headers is readable by all connectors between the sender and receiver. PSK payments use an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the inner PSK headers.
+
+Despite the fact that connectors can read the outer headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is end-to-end.
+
+The only headers in the outer headers should be the ones required for the receiver to decrypt the inner headers. In the above example, the decryption key for the outer headers' body is: `HMAC-SHA-256(data: "fpwpAhlN588", key: secret)`, where `secret` is the pre-shared secret key.
```

Comment:
That `fpwpAhlN588` should not be called data, because that would make me think of the application data the user supplies. That value should either be called the payment ID or nonce and we should refer to it consistently.